apps/common/lib/lexical/ast/analysis/analyzer.ex

```diff
@@ -0,0 +1,424 @@
+        else
+          {zipper, scopes}
+        end
+      end)
+
+    scopes
+  end
+
+  # extend range for block pairs to either the beginning of their next
+  # sibling or, if they are the last element, the end of their parent
+  defp maybe_correct_range(scope, %Zipper{node: {_, _}} = zipper, %Document{} = document) do
+    with %Zipper{node: sibling} <- Zipper.right(zipper),
+         %{start: sibling_start} <- Sourceror.get_range(sibling) do
+      new_end = Position.new(document, sibling_start[:line], sibling_start[:column])
+      put_in(scope.range.end, new_end)
+    else
+      _ ->
+        # we go up twice to get to the real parent because ast pairs
+        # are always in a list
+        %Zipper{node: parent} = zipper |> Zipper.up() |> Zipper.up()
+        %{end: parent_end} = Sourceror.get_range(parent)
+        new_end = Position.new(document, parent_end[:line], parent_end[:column])
+        put_in(scope.range.end, new_end)
+    end
+  end
+
+  defp maybe_correct_range(scope, _zipper, _document) do
+    scope
+  end
+
+  # add a unique ID to 3-element tuples
+  defp with_scope_id({_, _, _} = quoted) do
+    Macro.update_meta(quoted, &Keyword.put(&1, @scope_id, make_ref()))
+  end
+
+  defp with_scope_id(quoted) do
+    quoted
+  end
+
+  @doc false
+  def scope_id({_, meta, _}) when is_list(meta) do
+    Keyword.get(meta, @scope_id)
+  end
+
+  def scope_id({left, right}) do
+    {scope_id(left), scope_id(right)}
+  end
+
+  def scope_id(list) when is_list(list) do
+    Enum.map(list, &scope_id/1)
+  end
+
+  def scope_id(_) do
+    nil
+  end
+
+  # defmodule Foo do
+  defp pre({:defmodule, meta, [{:__aliases__, _, segments} | _]} = quoted, state) do
+    module =
+      case State.current_module(state) do
+        [] -> segments
+        current_module -> reify_alias(current_module, segments)
+      end
+
+    current_module_alias = Alias.new(module, :__MODULE__, meta[:line])
+
+    state
+    # implicit alias belongs to the current scope
+    |> maybe_push_implicit_alias(segments, meta[:line])
+    # new __MODULE__ alias belongs to the new scope
+    |> State.push_scope_for(quoted, module)
+    |> State.push_alias(current_module_alias)
+  end
+
+  # alias Foo.{Bar, Baz, Buzz.Qux}
+  defp pre({:alias, meta, [{{:., _, [aliases, :{}]}, _, aliases_nodes}]}, state) do
+    base_segments = expand_alias(aliases, state)
+
+    Enum.reduce(aliases_nodes, state, fn {:__aliases__, _, segments}, state ->
+      alias = Alias.new(base_segments ++ segments, List.last(segments), meta[:line])
+
+      state
+      |> State.push_alias(alias)
+    end)
+  end
+
+  # alias Foo
+  # alias Foo.Bar
+  # alias __MODULE__.Foo
+  defp pre({:alias, meta, [aliases]}, state) do
+    case expand_alias(aliases, state) do
+      [_ | _] = segments ->
+        alias = Alias.new(segments, List.last(segments), meta[:line])
+
+        state
+        |> State.push_alias(alias)
```

Comment:
i really, really, really  don't like single arg pipes
```suggestion
       State.push_alias(state, alias)
```

Reply:
Yep, a decent amount of this is somewhat leftover from an earlier refactoring where there were a couple of steps here instead of just one.